File: internal/prodgate/io_profile.go
package prodgate

// DbIoProfile holds driver-boundary SQL I/O counters for perf debugging.
type DbIoProfile struct {
	ConnectMS      int64 `json:"connect_ms,omitempty"`
	QueryMS        int64 `json:"query_ms"`
	QueryCalls     int64 `json:"query_calls"`
	FetchMS        int64 `json:"fetch_ms,omitempty"`
	FetchCalls     int64 `json:"fetch_calls,omitempty"`
	ExecMS         int64 `json:"exec_ms"`
	ExecCalls      int64 `json:"exec_calls"`
	ExtraConnects  int64 `json:"extra_connects,omitempty"`
	ExtraConnectMS int64 `json:"extra_connect_ms,omitempty"`
}

func (p DbIoProfile) DbBoundaryMS() int64 {
	return p.QueryMS + p.FetchMS + p.ExecMS
}

File: rust/crates/core/src/driver/io_profile.rs
use serde::{Deserialize, Serialize};

/// Driver-level SQL I/O counters (mirrors Go `timingConn` boundary stats).
#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(default)]
pub struct IoProfile {
    pub query_ms: i64,
    pub query_calls: i64,
    pub exec_ms: i64,
    pub exec_calls: i64,
    pub extra_connects: i64,
    pub extra_connect_ms: i64,
}

impl IoProfile {
    pub fn db_boundary_ms(&self) -> i64 {
        self.query_ms + self.exec_ms
    }
}
